src/Evolution/DiscontinuousGalerkin/InboxTags.hpp

```diff
@@ -0,0 +1,94 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/functional/hash.hpp>
+#include <cstddef>
+#include <map>
+#include <optional>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include "DataStructures/FixedHashMap.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Domain/Structure/MaxNumberOfNeighbors.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Parallel/InboxInserters.hpp"
+#include "Time/TimeStepId.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace evolution::dg::Tags {
+/// The inbox tag for boundary correction communication and DG-subcell ghost
+/// zone cells.
+///
+/// The first element in the tuple is the mesh of the neighboring element's
+/// face (not the mortar mesh!), the second is the volume variables for finite
+/// difference/volume reconstruction, and the third is the fluxes. The fourth
+/// element of the tuple is the range of validity of the boundary terms. This is
+/// need when using local time stepping so that the element knows whether or not
+/// it should remove boundary data and expect new data to be sent from a
+/// neighbor. The TimeStepId also coincides with the next time step of the
+/// neighboring element.
+template <size_t Dim>
+struct BoundaryCorrectionAndGhostCellsInbox {
+  using stored_type =
+      std::tuple<Mesh<Dim - 1>, std::optional<std::vector<double>>,
+                 std::optional<std::vector<double>>, ::TimeStepId>;
+
+ public:
+  using temporal_id = TimeStepId;
+  using type = std::map<
+      TimeStepId,
+      FixedHashMap<maximum_number_of_neighbors(Dim),
+                   std::pair<Direction<Dim>, ElementId<Dim>>, stored_type,
+                   boost::hash<std::pair<Direction<Dim>, ElementId<Dim>>>>>;
+
+  template <typename Inbox, typename ReceiveDataType>
+  static void insert_into_inbox(const gsl::not_null<Inbox*> inbox,
+                                const temporal_id& time_step_id,
+                                ReceiveDataType&& data) noexcept {
+    auto& current_inbox = (*inbox)[time_step_id];
+    // Since we could receive either flux data or ghost cell data, we need to be
+    // prepared for both.
+
+    if (auto it = current_inbox.find(data.first); it != current_inbox.end()) {
+      // We have already received either ghost cell or flux data at this time.
+      // Temporal ordering requires it be the ghost cells we have received.
```

Comment:
Please explain this claim, I have no idea what is going on.

Is this function called twice, first to send ghost cell data, then again to send fluxes?

Reply:
I have attempted to clarify this in the doxygen documentation. Again, please give me feedback on that so that hopefully this C++ comment can be used more to recall the dox comment :)